gusto/physics.py

```diff
@@ -78,31 +79,26 @@ def __init__(self, state):
         w_sat = (w_sat1 /
                  (p * exp(w_sat2 * (T - T_0) / (T - w_sat3)) - w_sat4))
 
-        # correct w_sat to be positive
-        w_sat = interpolate(conditional(w_sat > 0, w_sat, 999.0), V)
-
         # make appropriate condensation rate
-        cond_rate = ((self.water_v - w_sat) /
-                     (dt * (1.0 + ((L_v ** 2.0 * w_sat) /
-                                   cp * R_v * T ** 2.0))))
+        dot_r_cond = ((self.water_v - w_sat) /
+                      (dt * (1.0 + ((L_v ** 2.0 * w_sat) /
+                                    (cp * R_v * T ** 2.0)))))
 
         # adjust cond rate so negative concentrations don't occur
-        cond_rate = interpolate(conditional(cond_rate < 0,
-                                conditional(-cond_rate > self.water_c / dt,
-                                            -self.water_c / dt,
-                                            cond_rate),
-                                conditional(cond_rate > self.water_v / dt,
-                                            self.water_v / dt,
-                                            cond_rate)), V)
-
-        self.water_v_new = Interpolator(self.water_v - dt * cond_rate, V)
-        self.water_c_new = Interpolator(self.water_c + dt * cond_rate, V)
+        self.lim_cond_rate = Interpolator(conditional(dot_r_cond < 0,
+                                                      max_value(dot_r_cond, - self.water_c / dt),
+                                                      min_value(dot_r_cond, self.water_v / dt)), Vt)
+
+        self.cond_rate = Function(Vt)
+        self.water_v_new = Interpolator(self.water_v - dt * self.cond_rate, Vt)
+        self.water_c_new = Interpolator(self.water_c + dt * self.cond_rate, Vt)
         self.theta_new = Interpolator(self.theta *
-                                      (1.0 + dt * cond_rate *
+                                      (1.0 + dt * self.cond_rate *
                                        (cv * L_v / (c_vml * cp * T) -
-                                        R_v * cv * c_pml / (R_m * cp * c_vml))), V)
+                                        R_v * cv * c_pml / (R_m * cp * c_vml))), Vt)
 
     def apply(self):
+        self.cond_rate.assign(self.lim_cond_rate.interpolate())
         self.water_v.assign(self.water_v_new.interpolate())
         self.water_c.assign(self.water_c_new.interpolate())
         self.theta.assign(self.theta_new.interpolate())
```

Comment:
Oh, I see `theta` appears in the RHS of the theta expression, so probably not for that one, but cond_rate is purely diagnosed, so I don't think you need the intermediate variable?